evm_arithmetization/src/keccak_sponge/keccak_sponge_stark.rs

```diff
@@ -699,12 +753,23 @@ impl<F: RichField + Extendable<D>, const D: usize> Stark<F, D> for KeccakSpongeS
         );
         yield_constr.constraint(builder, constraint);
 
-        let is_final_block = builder.add_many_extension(local_values.is_final_input_len);
+        let is_final_block = local_values.is_padding_byte[KECCAK_RATE_BYTES - 1];
         let constraint = builder.mul_sub_extension(is_final_block, is_final_block, is_final_block);
         yield_constr.constraint(builder, constraint);
 
-        for &is_final_len in local_values.is_final_input_len.iter() {
-            let constraint = builder.mul_sub_extension(is_final_len, is_final_len, is_final_len);
+        for &is_padding_byte in local_values.is_padding_byte.iter() {
+            let constraint =
+                builder.mul_sub_extension(is_padding_byte, is_padding_byte, is_padding_byte);
+            yield_constr.constraint(builder, constraint);
+        }
```

Comment:
Same for circuit version, the first constraint (on `is_final_block`) is redundant.